### PR TITLE
Throw exception when reading from stream fails.

### PR DESCRIPTION
**Pull request by intrigus-lgtm**
#### Reference Issue

#3211 

#### What does this implement/fix? Explain your changes.
Fixes the problem that the code ignores failures when reading from a stream.
It would also have been possible to explicitly check the status of the stream after each read, but using exceptions seemed easier.

#### Any other comments?
There are probably other areas that have similar problems.
I only looked at this area, because it currently blocks https://github.com/google/oss-fuzz/pull/3856.
Note, I don't know whether I should have moved the new struct to a different position/file.
Feel free to move it.

### CODE REVIEW

File: Code/GraphMol/MolPickler.cpp [hunk 1]
```diff
@@ -801,55 +802,91 @@ void MolPickler::pickleMol(const ROMol &mol, std::string &ss) {
...
-    streamRead(ss, numAtoms, majorVersion);
-    if (numAtoms > 255) {
-      _depickle<int32_t>(ss, mol, majorVersion, numAtoms);
+  // Ensure that the exception mask of the `istream` is reset to the previous
+  // state after we're done.
+  IoExceptionStateResetter resetter(ss);
+  
+  // Enable exceptions, so we're notified when we've reached EOF or any other
+  // problem.
+  ss.exceptions(std::ifstream::eofbit | std::ifstream::failbit |
```
Comment (bp-kelley) [inline Code/GraphMol/MolPickler.cpp:832]: I like this, but I'd rather have the exception state passed into the IoExceptionStateRestter:

```
 IoExceptionStateResetter resetter(ss,
   std::ifstream::eofbit | std::ifstream::failbit | std::ifstream::badbit);
```

Which feels more RAII to me for some reason,

**Comment by bp-kelley** [comment]
@intrigus-lgtm, since we are here, should we do the streamWrite as well for symmetry?

**Comment by intrigus-lgtm** [comment]
> @intrigus-lgtm, since we are here, should we do the streamWrite as well for symmetry?

I don't know, but I did it anyway.

File: Code/GraphMol/MolPickler.cpp [hunk 2]
```diff
@@ -806,50 +849,74 @@ void MolPickler::pickleMol(const ROMol &mol, std::string &ss) {
 // will be blown out by the end of this process.
 void MolPickler::molFromPickle(std::istream &ss, ROMol *mol) {
   PRECONDITION(mol, "empty molecule");
-  int32_t tmpInt;
 
-  mol->clearAllAtomBookmarks();
-  mol->clearAllBondBookmarks();
+  // Ensure that the exception state of the `istream` is reset to the previous
```
Comment (greglandrum) [inline Code/GraphMol/MolPickler.cpp:853]: you mean `ostream` here

**Comment by bp-kelley** [comment]
I like it, @greglandrum what do you think?

**Comment by greglandrum** [comment]
LGTM!

**Pull request merged**